libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -471,6 +490,179 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
+    LOG(log_er_) << "Drop reward vote " << vote->getHash() << ". PBFT chain last block is "
+                 << pbft_chain_last_block_hash << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  if (vote->getType() != cert_vote_type) {
+    LOG(log_er_) << "Reward vote type is not cert vote type " << *vote;
+    return false;
+  }
+
+  if (vote->getStep() != 3) {
+    LOG(log_er_) << "Reward vote step is not 3 " << *vote;
+    return false;
+  }
+
+  uint64_t dpos_period = pbft_chain_->getPbftChainSize() - 1;  // reward period - 1
+  const auto& voter_account_addr = vote->getVoterAddr();
+  uint64_t voter_dpos_votes_count;
+  try {
+    voter_dpos_votes_count = final_chain_->dpos_eligible_vote_count(dpos_period, voter_account_addr);
+  } catch (state_api::ErrFutureBlock& c) {
+    LOG(log_er_) << c.what() << ". Voter account " << voter_account_addr << " in DPOS period " << dpos_period
+                 << " is too far for ahead of DPOS. Have executed chain size " << final_chain_->last_block_number();
+    return false;
+  }
+
+  if (!voter_dpos_votes_count) {
+    LOG(log_er_) << "Reward vote is invalid. DPOS votes count is 0 " << *vote;
+    return false;
+  }
+
+  const uint64_t reward_period_dpos_total_votes_count = final_chain_->dpos_eligible_total_vote_count(dpos_period);
+  const size_t reward_period_pbft_sortition_threshold =
+      db_->getPbftSortitionThreshold(dpos_period + 1);  // reward period
+  if (!reward_period_pbft_sortition_threshold) {
+    LOG(log_er_) << "Cannot get PBFT sortition threshold for period " << dpos_period;
+    return false;
+  }
+  try {
+    vote->validate(voter_dpos_votes_count, reward_period_dpos_total_votes_count,
+                   reward_period_pbft_sortition_threshold);
+  } catch (const std::logic_error& e) {
+    LOG(log_er_) << e.what();
+    LOG(log_er_) << "DPOS period " << dpos_period << ", voter " << voter_account_addr << " votes count "
+                 << voter_dpos_votes_count << ", total votes count " << reward_period_dpos_total_votes_count
+                 << ", pbft sortition threshold " << reward_period_pbft_sortition_threshold;
+
+    return false;
+  }
+
+  return true;
+}
+
+void VoteManager::updateRewardVotes(uint64_t reward_period) {
+  auto period_data = db_->getPeriodDataRaw(reward_period);
+  if (period_data.size() == 0) {
+    // First period no reward votes
+    return;
+  }
+
+  SyncBlock reward_period_sync_block(period_data);
+
+  std::unordered_set<vote_hash_t> reward_period_cert_votes_hash;
+  for (const auto& v : reward_period_sync_block.cert_votes) {
+    reward_period_cert_votes_hash.insert(v->getHash());
+  }
+
+  bool update = false;
+  {
+    const auto reward_period_pbft_block_hash = reward_period_sync_block.pbft_blk->getBlockHash();
+    std::unique_lock lock(reward_votes_mutex_);
+    if (reward_votes_.first != reward_period_pbft_block_hash) {
+      LOG(log_dg_) << "Not match reward period pbft block " << reward_period_pbft_block_hash
```

Comment:
Reward period pbft block hash <...> does not match ....

Reply:
Fixed